src/libraries/Common/src/System/Net/SocketAddress.cs

```diff
@@ -93,11 +93,32 @@ public SocketAddress(AddressFamily family, int size)
             }
 
             InternalSize = size;
-            Buffer = new byte[(size / IntPtr.Size + 2) * IntPtr.Size];
+
+            // TODO: workaround for https://github.com/dotnet/runtime/issues/33583
+            //       remove when Mono has pinned arrays
+            try
+            {
+                Buffer = GC.AllocateArray<byte>((size / IntPtr.Size + 2) * IntPtr.Size, pinned: true);
```

Comment:
@VSadov why does it make sense to use POH for SocketAddress.Buffer?

Reply:
My question is to better understand when it is better to use SOH vs POH for objects that will be pinned.

Reply:
In this change I basically look for arrays pinned via `GCHandleType.Pinned`. 
With few exceptions it is an indication that the object could be pinned for a nontrivial length of time, perhaps for its entire lifetime. `SocketAddress.Buffer` is one of such arrays.

mid/long term pinning in random locations could have really bad effects on Small Object Heap. 
On the other hand - if the object is expected to end up in Gen1/Gen2, then there is not much downside from allocating on POH.